2 error fixes with tests: "Add missing null check on the projection parameter." and "Fixed $pull never matching objects."

KCscores opened the pull request:
I have found that tingodb will throw an error on the following call when field `test1` does not exist in the matching document:
```
db.collection("test").find({_id: 1234}, {"test1.test2": 1}).toArray(function(err, result) {

});
```
If field `test1` exists but `test2` does not, the function will behave as expected - not provide the missing projection item.

With this change, both cases behave the same way.

sergeyksv commented:
Good catch :+1: Do you mind to add test for this fix please? Such special case tests are inside https://github.com/sergeyksv/tingodb/blob/master/test/misc-test.js
Will appreciate if you make it.

KCscores commented:
I apologize for the long delay. I ran into another issue with tingodb, fixed it, wrote a test, and swung by to write a test for the original issue as well. I've renamed this request accordingly to now encompass both issues.

The new issue is thus: $pull never matches objects. I found that when you pass an object as a parameter to pull:
```
var projection = { $pull: {
   arr: {
      name: "Bob",
      age: "10"
   }
}};
db.collection("coll").update({ _id: 1 }, projection, {multi: true, upsert: true}, function(err, result) {});
```
This will never, ever match internally because when the object is passed to to matcher, it is always inserted inside an object `v`. When the matcher is created and you match a non-object with $pull (ex. number) the matcher puts it inside an object `v`, but when you match an object, it does not. This means that objects passed to the matcher need to be inserted as-is and not put into `v` first.

sergeyksv commented:
Thanks :+1:
